### PR TITLE
contrib/labstack/echo: Add Labstack Echo middleware

### LINKED ISSUE #174: proposal: add support for Labstack Echo
Hi!

It would be nice to have an implementation for Labstack's Echo web framework.

### PR DESCRIPTION
**Pull request by rafaeljusto**
Continue work from #383, applying suggestions from @gbbr
Credits to @HatsuneMiku3939

Closes #174

### CODE REVIEW

**Comment by gbbr** [comment]
Thanks @rafaeljusto. We're going to review this soon.

File: contrib/labstack/echo/option.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package echo
+
+type config struct {
+	serviceName string
+}
+
+// MiddlewareOption represents an option that can be passed to Middleware.
+type MiddlewareOption func(*config)
```
Comment (gbbr) [inline contrib/labstack/echo/option.go]: Can we please call this `Option`?

File: contrib/labstack/echo/echotrace_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	assert.Equal("GET", span.Tag(ext.HTTPMethod))
+
+	assert.Equal("/user/123", span.Tag(ext.HTTPURL))
+}
+
+func TestError(t *testing.T) {
+	assert := assert.New(t)
+	mt := mocktracer.Start()
+	defer mt.Stop()
+	var called bool
```
Comment (gbbr) [inline contrib/labstack/echo/echotrace_test.go]: We can simplify here:
```go
var called, traced bool
```
And in all other tests.

File: contrib/labstack/echo/echotrace.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			c.SetRequest(request.WithContext(ctx))
+
+			// serve the request to the next middleware
+			err := next(c)
+
+			span.SetTag(ext.HTTPCode, strconv.Itoa(c.Response().Status))
+			if err != nil {
+				span.SetTag(ext.Error, err)
+			}
+
```
Comment (gbbr) [inline contrib/labstack/echo/echotrace.go]: Can you please remove this empty line?

File: contrib/labstack/echo/echotrace.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+// Middleware returns echo middleware which will trace incoming requests.
+func Middleware(opts ...MiddlewareOption) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		cfg := new(config)
+		defaults(cfg)
+		for _, fn := range opts {
+			fn(cfg)
+		}
+
```
Comment (gbbr) [inline contrib/labstack/echo/echotrace.go]: Can you please remove this empty line too?

**Comment by rafaeljusto** [comment]
Applied the changes there @gbbr 👍

**Comment by rafaeljusto** [comment]
Tests failed with a panic in `contrib/gocql/gocql` package 🤔 
```
=== RUN   TestErrorWrapper
--- FAIL: TestErrorWrapper (0.60s)
    gocql_test.go:62: 
        	Error Trace:	gocql_test.go:62
        	Error:      	Expected nil, but got: &errors.errorString{s:"gocql: unable to create session: unable to discover protocol version: gocql: no response to connection startup within timeout"}
        	Test:       	TestErrorWrapper
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8ad51d]

goroutine 55 [running]:
testing.tRunner.func1(0xc000286300)
	/usr/local/go/src/testing/testing.go:830 +0x69d
panic(0x942ac0, 0xd29be0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/gocql/gocql.(*Query).defaultsFromSession(0xc00027a360)
	/go/src/github.com/gocql/gocql/session.go:700 +0x5d
github.com/gocql/gocql.(*Session).Query(0x0, 0x9c185c, 0x64, 0x0, 0x0, 0x0, 0x23c34600)
	/go/src/github.com/gocql/gocql/session.go:336 +0x10f
gopkg.in/DataDog/dd-trace-go.v1/contrib/gocql/gocql.TestErrorWrapper(0xc000286300)
	/go/src/gopkg.in/DataDog/dd-trace-go.v1/contrib/gocql/gocql/gocql_test.go:63 +0x430
testing.tRunner(0xc000286300, 0x9c5ef8)
	/usr/local/go/src/testing/testing.go:865 +0x164
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x65b
FAIL	gopkg.in/DataDog/dd-trace-go.v1/contrib/gocql/gocql	0.951s
```

Non-deterministic test? Maybe running it again it will work. 😄 

Btw, this new echo contribution is not 100% test coverage. 😮 
image

Any suggestions for a test case to cover this scenario? Or should we leave it as is?

**Comment by gbbr** [comment]
That's probably a flake, don't worry about it. About the test coverage, great question! All you have to do is create a root span in the test and then inject it into the headers of the request you make, for example:
```go
root := tracer.StartSpan("root")
r := httptest.NewRequest("GET", "/ping", nil)
err := tracer.Inject(root.Context(), tracer.HTTPHeadersCarrier(r.Header))
assert.Nil(err)
// ...
```
And then later on we should ensure that the span the integration created is a child of this root.

**Comment by rafaeljusto** [comment]
Thanks @gbbr ! Now we got 100% test coverage 🎉

**Comment by gbbr** [comment]
This is great! Thanks a lot! It took a while to get this merged 😄

**Comment by rafaeljusto** [comment]
Hey @gbbr , is it possible to create a new tag (maybe `v1.14.0`) that includes this feature?

**Comment by gbbr** [comment]
We will do that soon.

**Comment by gbbr** [comment]
This has been released as part of https://github.com/DataDog/dd-trace-go/releases/tag/v1.14.0

**Comment by rafaeljusto** [comment]
Thanks @gbbr ! 😄

**Pull request merged**